File: code/packages/rust/graph/src/graph.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fmt;

use crate::algorithms::TraversalGraph;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GraphRepr {
    AdjacencyList,
    AdjacencyMatrix,
}

pub type WeightedEdge = (String, String, f64);

#[derive(Debug, Clone, PartialEq)]
pub enum GraphPropertyValue {
    String(String),
    Number(f64),
    Bool(bool),
    Null,
}

pub type GraphPropertyBag = BTreeMap<String, GraphPropertyValue>;

#[derive(Debug, Clone, PartialEq)]
pub enum GraphError {
    NodeNotFound(String),
    EdgeNotFound(String, String),
    NotConnected,
}

impl fmt::Display for GraphError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GraphError::NodeNotFound(node) => write!(f, "node not found: {}", node),
            GraphError::EdgeNotFound(left, right) => {
                write!(f, "edge not found: {} -- {}", left, right)
            }
            GraphError::NotConnected => write!(f, "graph is not connected"),
        }
    }
}

impl std::error::Error for GraphError {}

#[derive(Debug, Clone)]
pub struct Graph {
    repr: GraphRepr,
    adj: BTreeMap<String, BTreeMap<String, f64>>,
    node_list: Vec<String>,
    node_index: BTreeMap<String, usize>,
    matrix: Vec<Vec<Option<f64>>>,
    graph_properties: GraphPropertyBag,
    node_properties: BTreeMap<String, GraphPropertyBag>,
    edge_properties: BTreeMap<(String, String), GraphPropertyBag>,
}

impl Graph {
    pub fn new(repr: GraphRepr) -> Self {
        Self {
            repr,
            adj: BTreeMap::new(),
            node_list: Vec::new(),
            node_index: BTreeMap::new(),
            matrix: Vec::new(),
            graph_properties: BTreeMap::new(),
            node_properties: BTreeMap::new(),
            edge_properties: BTreeMap::new(),
        }
    }

    pub fn repr(&self) -> GraphRepr {
        self.repr
    }

    pub fn add_node(&mut self, node: impl Into<String>) {
        let node = node.into();
        self.add_node_with_properties(node, BTreeMap::new());
    }

    pub fn add_node_with_properties(
        &mut self,
        node: impl Into<String>,
        properties: GraphPropertyBag,
    ) {
        let node = node.into();
        match self.repr {
            GraphRepr::AdjacencyList => {
                self.adj.entry(node.clone()).or_default();
            }
            GraphRepr::AdjacencyMatrix => {
                if !self.node_index.contains_key(&node) {
                    let index = self.node_list.len();
                    self.node_list.push(node.clone());
                    self.node_index.insert(node.clone(), index);
                    for row in &mut self.matrix {
                        row.push(None);
                    }
                    self.matrix.push(vec![None; index + 1]);
                }
            }
        }
        self.node_properties
            .entry(node)
            .or_default()
            .extend(properties);
    }

    pub fn remove_node(&mut self, node: &str) -> Result<(), GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => {
                let neighbors = self
                    .adj
                    .get(node)
                    .cloned()
                    .ok_or_else(|| GraphError::NodeNotFound(node.to_string()))?;
                for neighbor in neighbors.keys() {
                    if let Some(entry) = self.adj.get_mut(neighbor) {
                        entry.remove(node);
                    }
                    self.edge_properties
                        .remove(&canonical_endpoints(node, neighbor));
                }
                self.adj.remove(node);
                self.node_properties.remove(node);
                Ok(())
            }
            GraphRepr::AdjacencyMatrix => {
                let index = self
                    .node_index
                    .remove(node)
                    .ok_or_else(|| GraphError::NodeNotFound(node.to_string()))?;
                for other in &self.node_list {
                    self.edge_properties
                        .remove(&canonical_endpoints(node, other));
                }
                self.node_list.remove(index);
                self.node_properties.remove(node);
                self.matrix.remove(index);
                for row in &mut self.matrix {
                    row.remove(index);
                }
                for (offset, name) in self.node_list[index..].iter().enumerate() {
                    self.node_index.insert(name.clone(), index + offset);
                }
                Ok(())
            }
        }
    }

    pub fn has_node(&self, node: &str) -> bool {
        match self.repr {
            GraphRepr::AdjacencyList => self.adj.contains_key(node),
            GraphRepr::AdjacencyMatrix => self.node_index.contains_key(node),
        }
    }

    pub fn nodes(&self) -> Vec<String> {
        let mut nodes = match self.repr {
            GraphRepr::AdjacencyList => self.adj.keys().cloned().collect(),
            GraphRepr::AdjacencyMatrix => self.node_list.clone(),
        };
        nodes.sort();
        nodes
    }

    pub fn add_edge(&mut self, left: impl Into<String>, right: impl Into<String>, weight: f64) {
        self.add_edge_with_properties(left, right, weight, BTreeMap::new());
    }

    pub fn add_edge_with_properties(
        &mut self,
        left: impl Into<String>,
        right: impl Into<String>,
        weight: f64,
        properties: GraphPropertyBag,
    ) {
        let left = left.into();
        let right = right.into();
        self.add_node(left.clone());
        self.add_node(right.clone());

        match self.repr {
            GraphRepr::AdjacencyList => {
                self.adj
                    .get_mut(&left)
                    .expect("left node must exist")
                    .insert(right.clone(), weight);
                self.adj
                    .get_mut(&right)
                    .expect("right node must exist")
                    .insert(left.clone(), weight);
            }
            GraphRepr::AdjacencyMatrix => {
                let left_index = self.node_index[&left];
                let right_index = self.node_index[&right];
                self.matrix[left_index][right_index] = Some(weight);
                self.matrix[right_index][left_index] = Some(weight);
            }
        }
        let edge_key = canonical_endpoints(&left, &right);
        let edge_properties = self.edge_properties.entry(edge_key).or_default();
        edge_properties.extend(properties);
        edge_properties.insert("weight".to_string(), GraphPropertyValue::Number(weight));
    }

    pub fn remove_edge(&mut self, left: &str, right: &str) -> Result<(), GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => {
                if !self
                    .adj
                    .get(left)
                    .map(|neighbors| neighbors.contains_key(right))
                    .unwrap_or(false)
                {
                    return Err(GraphError::EdgeNotFound(
                        left.to_string(),
                        right.to_string(),
                    ));
                }
                self.adj.get_mut(left).unwrap().remove(right);
                self.adj.get_mut(right).unwrap().remove(left);
                self.edge_properties
                    .remove(&canonical_endpoints(left, right));
                Ok(())
            }
            GraphRepr::AdjacencyMatrix => {
                let left_index =
                    self.node_index.get(left).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                let right_index =
                    self.node_index.get(right).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                if self.matrix[left_index][right_index].is_none() {
                    return Err(GraphError::EdgeNotFound(
                        left.to_string(),
                        right.to_string(),
                    ));
                }
                self.matrix[left_index][right_index] = None;
                self.matrix[right_index][left_index] = None;
                self.edge_properties
                    .remove(&canonical_endpoints(left, right));
                Ok(())
            }
        }
    }

    pub fn has_edge(&self, left: &str, right: &str) -> bool {
        match self.repr {
            GraphRepr::AdjacencyList => self
                .adj
                .get(left)
                .map(|neighbors| neighbors.contains_key(right))
                .unwrap_or(false),
            GraphRepr::AdjacencyMatrix => self
                .node_index
                .get(left)
                .zip(self.node_index.get(right))
                .and_then(|(left_index, right_index)| self.matrix[*left_index][*right_index])
                .is_some(),
        }
    }

    pub fn edge_weight(&self, left: &str, right: &str) -> Result<f64, GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => self
                .adj
                .get(left)
                .and_then(|neighbors| neighbors.get(right))
                .copied()
                .ok_or_else(|| GraphError::EdgeNotFound(left.to_string(), right.to_string())),
            GraphRepr::AdjacencyMatrix => {
                let left_index =
                    self.node_index.get(left).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                let right_index =
                    self.node_index.get(right).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                self.matrix[left_index][right_index]
                    .ok_or_else(|| GraphError::EdgeNotFound(left.to_string(), right.to_string()))
            }
        }
    }

    pub fn graph_properties(&self) -> GraphPropertyBag {
        self.graph_properties.clone()
    }

    pub fn set_graph_property(&mut self, key: impl Into<String>, value: GraphPropertyValue) {
        self.graph_properties.insert(key.into(), value);
    }

    pub fn remove_graph_property(&mut self, key: &str) {
        self.graph_properties.remove(key);
    }

    pub fn node_properties(&self, node: &str) -> Result<GraphPropertyBag, GraphError> {
        if !self.has_node(node) {
            return Err(GraphError::NodeNotFound(node.to_string()));
        }
        Ok(self.node_properties.get(node).cloned().unwrap_or_default())
    }

    pub fn set_node_property(
        &mut self,
        node: &str,
        key: impl Into<String>,
        value: GraphPropertyValue,
    ) -> Result<(), GraphError> {
        if !self.has_node(node) {
            return Err(GraphError::NodeNotFound(node.to_string()));
        }
        self.node_properties
            .entry(node.to_string())
            .or_default()
            .insert(key.into(), value);
        Ok(())
    }

    pub fn remove_node_property(&mut self, node: &str, key: &str) -> Result<(), GraphError> {
        if !self.has_node(node) {
            return Err(GraphError::NodeNotFound(node.to_string()));
        }
        if let Some(properties) = self.node_properties.get_mut(node) {
            properties.remove(key);
        }
        Ok(())
    }

    pub fn edge_properties(&self, left: &str, right: &str) -> Result<GraphPropertyBag, GraphError> {
        if !self.has_edge(left, right) {
            return Err(GraphError::EdgeNotFound(
                left.to_string(),
                right.to_string(),
            ));
        }
        let mut properties = self
            .edge_properties
            .get(&canonical_endpoints(left, right))
            .cloned()
            .unwrap_or_default();
        properties.insert(
            "weight".to_string(),
            GraphPropertyValue::Number(self.edge_weight(left, right)?),
        );
        Ok(properties)
    }

    pub fn set_edge_property(
        &mut self,
        left: &str,
        right: &str,
        key: impl Into<String>,
        value: GraphPropertyValue,
    ) -> Result<(), GraphError> {
        if !self.has_edge(left, right) {
            return Err(GraphError::EdgeNotFound(
                left.to_string(),
                right.to_string(),
            ));
        }
        let key = key.into();
        if key == "weight" {
            match value {
                GraphPropertyValue::Number(weight) => {
                    self.set_edge_weight(left, right, weight)?;
                    self.edge_properties
                        .entry(canonical_endpoints(left, right))
                        .or_default()
                        .insert(key, GraphPropertyValue::Number(weight));
                    return Ok(());
                }
                _ => {
                    return Err(GraphError::EdgeNotFound(
                        "weight".to_string(),
                        "numeric property".to_string(),
                    ));
                }
            }
        }
        self.edge_properties
            .entry(canonical_endpoints(left, right))
            .or_default()
            .insert(key, value);
        Ok(())
    }

    pub fn remove_edge_property(
        &mut self,
        left: &str,
        right: &str,
        key: &str,
    ) -> Result<(), GraphError> {
        if !self.has_edge(left, right) {
            return Err(GraphError::EdgeNotFound(
                left.to_string(),
                right.to_string(),
            ));
        }
        if key == "weight" {
            self.set_edge_weight(left, right, 1.0)?;
            self.edge_properties
                .entry(canonical_endpoints(left, right))
                .or_default()
                .insert("weight".to_string(), GraphPropertyValue::Number(1.0));
            return Ok(());
        }
        if let Some(properties) = self
            .edge_properties
            .get_mut(&canonical_endpoints(left, right))
        {
            properties.remove(key);
        }
        Ok(())
    }

    pub fn edges(&self) -> Vec<WeightedEdge> {
        let mut result = Vec::new();
        match self.repr {
            GraphRepr::AdjacencyList => {
                let mut seen = BTreeSet::new();
                for (left, neighbors) in &self.adj {
                    for (right, weight) in neighbors {
                        let (first, second) = canonical_endpoints(left, right);
                        if seen.insert((first.clone(), second.clone())) {
                            result.push((first, second, *weight));
                        }
                    }
                }
            }
            GraphRepr::AdjacencyMatrix => {
                for row in 0..self.node_list.len() {
                    for col in row..self.node_list.len() {
                        if let Some(weight) = self.matrix[row][col] {
                            result.push((
                                self.node_list[row].clone(),
                                self.node_list[col].clone(),
                                weight,
                            ));
                        }
                    }
                }
            }
        }
        result.sort_by(|left, right| {
            left.2
                .total_cmp(&right.2)
                .then_with(|| left.0.cmp(&right.0))
                .then_with(|| left.1.cmp(&right.1))
        });
        result
    }

    pub fn neighbors(&self, node: &str) -> Result<Vec<String>, GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => self
                .adj
                .get(node)
                .map(|neighbors| neighbors.keys().cloned().collect())
                .ok_or_else(|| GraphError::NodeNotFound(node.to_string())),
            GraphRepr::AdjacencyMatrix => {
                let index = self
                    .node_index
                    .get(node)
                    .copied()
                    .ok_or_else(|| GraphError::NodeNotFound(node.to_string()))?;
                let mut neighbors: Vec<String> = self.matrix[index]
                    .iter()
                    .enumerate()
                    .filter_map(|(col, weight)| weight.map(|_| self.node_list[col].clone()))
                    .collect();
                neighbors.sort();
                Ok(neighbors)
            }
        }
    }

    pub fn neighbors_weighted(&self, node: &str) -> Result<BTreeMap<String, f64>, GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => self
                .adj
                .get(node)
                .cloned()
                .ok_or_else(|| GraphError::NodeNotFound(node.to_string())),
            GraphRepr::AdjacencyMatrix => {
                let index = self
                    .node_index
                    .get(node)
                    .copied()
                    .ok_or_else(|| GraphError::NodeNotFound(node.to_string()))?;
                let mut result = BTreeMap::new();
                for (col, weight) in self.matrix[index].iter().enumerate() {
                    if let Some(weight) = weight {
                        result.insert(self.node_list[col].clone(), *weight);
                    }
                }
                Ok(result)
            }
        }
    }

    pub fn degree(&self, node: &str) -> Result<usize, GraphError> {
        Ok(self.neighbors(node)?.len())
    }

    pub fn size(&self) -> usize {
        match self.repr {
            GraphRepr::AdjacencyList => self.adj.len(),
            GraphRepr::AdjacencyMatrix => self.node_list.len(),
        }
    }

    fn set_edge_weight(&mut self, left: &str, right: &str, weight: f64) -> Result<(), GraphError> {
        match self.repr {
            GraphRepr::AdjacencyList => {
                if !self.has_edge(left, right) {
                    return Err(GraphError::EdgeNotFound(
                        left.to_string(),
                        right.to_string(),
                    ));
                }
                self.adj
                    .get_mut(left)
                    .expect("left node must exist")
                    .insert(right.to_string(), weight);
                self.adj
                    .get_mut(right)
                    .expect("right node must exist")
                    .insert(left.to_string(), weight);
            }
            GraphRepr::AdjacencyMatrix => {
                let left_index =
                    self.node_index.get(left).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                let right_index =
                    self.node_index.get(right).copied().ok_or_else(|| {
                        GraphError::EdgeNotFound(left.to_string(), right.to_string())
                    })?;
                if self.matrix[left_index][right_index].is_none() {
                    return Err(GraphError::EdgeNotFound(
                        left.to_string(),
                        right.to_string(),
                    ));
                }
                self.matrix[left_index][right_index] = Some(weight);
                self.matrix[right_index][left_index] = Some(weight);
            }
        }
        Ok(())
    }
}

impl Default for Graph {
    fn default() -> Self {
        Self::new(GraphRepr::AdjacencyList)
    }
}

impl fmt::Display for Graph {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Graph(nodes={}, edges={}, repr={:?})",
            self.size(),
            self.edges().len(),
            self.repr
        )
    }
}

impl TraversalGraph for Graph {
    type Error = GraphError;

    fn has_node(&self, node: &str) -> bool {
        Graph::has_node(self, node)
    }

    fn neighbors(&self, node: &str) -> Result<Vec<String>, Self::Error> {
        Graph::neighbors(self, node)
    }
}

fn canonical_endpoints(left: &str, right: &str) -> (String, String) {
    if left <= right {
        (left.to_string(), right.to_string())
    } else {
        (right.to_string(), left.to_string())
    }
}

File: code/packages/go/graph/graph.go
package graph

import (
	"fmt"
	"sort"
)

type GraphRepr string

const (
	AdjacencyList   GraphRepr = "adjacency_list"
	AdjacencyMatrix GraphRepr = "adjacency_matrix"
)

type WeightedEdge struct {
	Left   string
	Right  string
	Weight float64
}

type PropertyValue any
type PropertyBag map[string]PropertyValue

type Graph struct {
	repr            GraphRepr
	adj             map[string]map[string]float64
	nodeList        []string
	nodeIndex       map[string]int
	matrix          [][]float64
	graphProperties PropertyBag
	nodeProperties  map[string]PropertyBag
	edgeProperties  map[string]PropertyBag
}

func New(repr GraphRepr) *Graph {
	if repr == "" {
		repr = AdjacencyList
	}
	return &Graph{
		repr:            repr,
		adj:             make(map[string]map[string]float64),
		nodeList:        []string{},
		nodeIndex:       make(map[string]int),
		matrix:          [][]float64{},
		graphProperties: make(PropertyBag),
		nodeProperties:  make(map[string]PropertyBag),
		edgeProperties:  make(map[string]PropertyBag),
	}
}

func (g *Graph) Repr() GraphRepr {
	return g.repr
}

func (g *Graph) AddNode(node string, properties ...PropertyBag) {
	if g.repr == AdjacencyList {
		if _, ok := g.adj[node]; !ok {
			g.adj[node] = make(map[string]float64)
			g.nodeProperties[node] = make(PropertyBag)
		}
		g.mergeNodeProperties(node, properties...)
		return
	}

	if _, ok := g.nodeIndex[node]; ok {
		g.mergeNodeProperties(node, properties...)
		return
	}

	index := len(g.nodeList)
	g.nodeList = append(g.nodeList, node)
	g.nodeIndex[node] = index
	g.nodeProperties[node] = make(PropertyBag)
	g.mergeNodeProperties(node, properties...)
	for i := range g.matrix {
		g.matrix[i] = append(g.matrix[i], 0)
	}
	g.matrix = append(g.matrix, make([]float64, index+1))
}

func (g *Graph) RemoveNode(node string) error {
	if g.repr == AdjacencyList {
		neighbors, ok := g.adj[node]
		if !ok {
			return &NodeNotFoundError{Node: node}
		}
		for neighbor := range neighbors {
			delete(g.adj[neighbor], node)
			delete(g.edgeProperties, edgeKey(node, neighbor))
		}
		delete(g.adj, node)
		delete(g.nodeProperties, node)
		return nil
	}

	index, ok := g.nodeIndex[node]
	if !ok {
		return &NodeNotFoundError{Node: node}
	}
	for _, other := range g.nodeList {
		delete(g.edgeProperties, edgeKey(node, other))
	}

	delete(g.nodeIndex, node)
	delete(g.nodeProperties, node)
	g.nodeList = append(g.nodeList[:index], g.nodeList[index+1:]...)
	g.matrix = append(g.matrix[:index], g.matrix[index+1:]...)
	for i := range g.matrix {
		g.matrix[i] = append(g.matrix[i][:index], g.matrix[i][index+1:]...)
	}
	for i := index; i < len(g.nodeList); i++ {
		g.nodeIndex[g.nodeList[i]] = i
	}
	return nil
}

func (g *Graph) HasNode(node string) bool {
	if g.repr == AdjacencyList {
		_, ok := g.adj[node]
		return ok
	}
	_, ok := g.nodeIndex[node]
	return ok
}

func (g *Graph) Nodes() []string {
	if g.repr == AdjacencyList {
		nodes := make([]string, 0, len(g.adj))
		for node := range g.adj {
			nodes = append(nodes, node)
		}
		sort.Strings(nodes)
		return nodes
	}
	return append([]string(nil), g.nodeList...)
}

func (g *Graph) AddEdge(left, right string, weight float64, properties ...PropertyBag) {
	if weight == 0 {
		weight = 1
	}
	g.AddNode(left)
	g.AddNode(right)

	if g.repr == AdjacencyList {
		g.adj[left][right] = weight
		g.adj[right][left] = weight
		g.mergeEdgeProperties(left, right, weight, properties...)
		return
	}

	leftIndex := g.nodeIndex[left]
	rightIndex := g.nodeIndex[right]
	g.matrix[leftIndex][rightIndex] = weight
	g.matrix[rightIndex][leftIndex] = weight
	g.mergeEdgeProperties(left, right, weight, properties...)
}

func (g *Graph) RemoveEdge(left, right string) error {
	if g.repr == AdjacencyList {
		neighbors, ok := g.adj[left]
		if !ok {
			return &EdgeNotFoundError{Left: left, Right: right}
		}
		if _, ok := neighbors[right]; !ok {
			return &EdgeNotFoundError{Left: left, Right: right}
		}
		delete(g.adj[left], right)
		delete(g.adj[right], left)
		delete(g.edgeProperties, edgeKey(left, right))
		return nil
	}

	leftIndex, okLeft := g.nodeIndex[left]
	rightIndex, okRight := g.nodeIndex[right]
	if !okLeft || !okRight || g.matrix[leftIndex][rightIndex] == 0 {
		return &EdgeNotFoundError{Left: left, Right: right}
	}
	g.matrix[leftIndex][rightIndex] = 0
	g.matrix[rightIndex][leftIndex] = 0
	delete(g.edgeProperties, edgeKey(left, right))
	return nil
}

func (g *Graph) HasEdge(left, right string) bool {
	if g.repr == AdjacencyList {
		if neighbors, ok := g.adj[left]; ok {
			_, ok = neighbors[right]
			return ok
		}
		return false
	}

	leftIndex, okLeft := g.nodeIndex[left]
	rightIndex, okRight := g.nodeIndex[right]
	if !okLeft || !okRight {
		return false
	}
	return g.matrix[leftIndex][rightIndex] != 0
}

func (g *Graph) EdgeWeight(left, right string) (float64, error) {
	if g.repr == AdjacencyList {
		if neighbors, ok := g.adj[left]; ok {
			if weight, ok := neighbors[right]; ok {
				return weight, nil
			}
		}
		return 0, &EdgeNotFoundError{Left: left, Right: right}
	}

	leftIndex, okLeft := g.nodeIndex[left]
	rightIndex, okRight := g.nodeIndex[right]
	if !okLeft || !okRight || g.matrix[leftIndex][rightIndex] == 0 {
		return 0, &EdgeNotFoundError{Left: left, Right: right}
	}
	return g.matrix[leftIndex][rightIndex], nil
}

func (g *Graph) Edges() []WeightedEdge {
	result := make([]WeightedEdge, 0)
	seen := make(map[string]bool)

	if g.repr == AdjacencyList {
		for left, neighbors := range g.adj {
			for right, weight := range neighbors {
				keyLeft, keyRight := canonicalEndpoints(left, right)
				key := keyLeft + "\x00" + keyRight
				if seen[key] {
					continue
				}
				seen[key] = true
				result = append(result, WeightedEdge{Left: keyLeft, Right: keyRight, Weight: weight})
			}
		}
	} else {
		for row := 0; row < len(g.nodeList); row++ {
			for col := row; col < len(g.nodeList); col++ {
				if g.matrix[row][col] != 0 {
					result = append(result, WeightedEdge{
						Left:   g.nodeList[row],
						Right:  g.nodeList[col],
						Weight: g.matrix[row][col],
					})
				}
			}
		}
	}

	sort.Slice(result, func(i, j int) bool {
		if result[i].Weight != result[j].Weight {
			return result[i].Weight < result[j].Weight
		}
		if result[i].Left != result[j].Left {
			return result[i].Left < result[j].Left
		}
		return result[i].Right < result[j].Right
	})
	return result
}

func (g *Graph) Neighbors(node string) ([]string, error) {
	if g.repr == AdjacencyList {
		neighbors, ok := g.adj[node]
		if !ok {
			return nil, &NodeNotFoundError{Node: node}
		}
		result := make([]string, 0, len(neighbors))
		for neighbor := range neighbors {
			result = append(result, neighbor)
		}
		sort.Strings(result)
		return result, nil
	}

	index, ok := g.nodeIndex[node]
	if !ok {
		return nil, &NodeNotFoundError{Node: node}
	}
	result := make([]string, 0)
	for col, weight := range g.matrix[index] {
		if weight != 0 {
			result = append(result, g.nodeList[col])
		}
	}
	sort.Strings(result)
	return result, nil
}

func (g *Graph) NeighborsWeighted(node string) (map[string]float64, error) {
	if g.repr == AdjacencyList {
		neighbors, ok := g.adj[node]
		if !ok {
			return nil, &NodeNotFoundError{Node: node}
		}
		result := make(map[string]float64, len(neighbors))
		for neighbor, weight := range neighbors {
			result[neighbor] = weight
		}
		return result, nil
	}

	index, ok := g.nodeIndex[node]
	if !ok {
		return nil, &NodeNotFoundError{Node: node}
	}
	result := make(map[string]float64)
	for col, weight := range g.matrix[index] {
		if weight != 0 {
			result[g.nodeList[col]] = weight
		}
	}
	return result, nil
}

func (g *Graph) Degree(node string) (int, error) {
	neighbors, err := g.Neighbors(node)
	if err != nil {
		return 0, err
	}
	return len(neighbors), nil
}

func (g *Graph) GraphProperties() PropertyBag {
	return copyPropertyBag(g.graphProperties)
}

func (g *Graph) SetGraphProperty(key string, value PropertyValue) {
	g.graphProperties[key] = value
}

func (g *Graph) RemoveGraphProperty(key string) {
	delete(g.graphProperties, key)
}

func (g *Graph) NodeProperties(node string) (PropertyBag, error) {
	if !g.HasNode(node) {
		return nil, &NodeNotFoundError{Node: node}
	}
	return copyPropertyBag(g.nodeProperties[node]), nil
}

func (g *Graph) SetNodeProperty(node, key string, value PropertyValue) error {
	if !g.HasNode(node) {
		return &NodeNotFoundError{Node: node}
	}
	g.nodeProperties[node][key] = value
	return nil
}

func (g *Graph) RemoveNodeProperty(node, key string) error {
	if !g.HasNode(node) {
		return &NodeNotFoundError{Node: node}
	}
	delete(g.nodeProperties[node], key)
	return nil
}

func (g *Graph) EdgeProperties(left, right string) (PropertyBag, error) {
	if !g.HasEdge(left, right) {
		return nil, &EdgeNotFoundError{Left: left, Right: right}
	}
	properties := copyPropertyBag(g.edgeProperties[edgeKey(left, right)])
	weight, err := g.EdgeWeight(left, right)
	if err != nil {
		return nil, err
	}
	properties["weight"] = weight
	return properties, nil
}

func (g *Graph) SetEdgeProperty(left, right, propertyKey string, value PropertyValue) error {
	if !g.HasEdge(left, right) {
		return &EdgeNotFoundError{Left: left, Right: right}
	}
	if propertyKey == "weight" {
		weight, ok := numericValue(value)
		if !ok {
			return fmt.Errorf("edge property %q must be numeric", propertyKey)
		}
		g.setEdgeWeight(left, right, weight)
	}
	edgeID := edgeKey(left, right)
	if _, ok := g.edgeProperties[edgeID]; !ok {
		g.edgeProperties[edgeID] = make(PropertyBag)
	}
	g.edgeProperties[edgeID][propertyKey] = value
	return nil
}

func (g *Graph) RemoveEdgeProperty(left, right, key string) error {
	if !g.HasEdge(left, right) {
		return &EdgeNotFoundError{Left: left, Right: right}
	}
	if key == "weight" {
		g.setEdgeWeight(left, right, 1)
		edgeID := edgeKey(left, right)
		if _, ok := g.edgeProperties[edgeID]; !ok {
			g.edgeProperties[edgeID] = make(PropertyBag)
		}
		g.edgeProperties[edgeID]["weight"] = float64(1)
		return nil
	}
	delete(g.edgeProperties[edgeKey(left, right)], key)
	return nil
}

func (g *Graph) Size() int {
	if g.repr == AdjacencyList {
		return len(g.adj)
	}
	return len(g.nodeList)
}

func (g *Graph) String() string {
	return fmt.Sprintf("Graph(nodes=%d, edges=%d, repr=%s)", g.Size(), len(g.Edges()), g.repr)
}

func canonicalEndpoints(left, right string) (string, string) {
	if left <= right {
		return left, right
	}
	return right, left
}

func edgeKey(left, right string) string {
	keyLeft, keyRight := canonicalEndpoints(left, right)
	return keyLeft + "\x00" + keyRight
}

func copyPropertyBag(properties PropertyBag) PropertyBag {
	result := make(PropertyBag, len(properties))
	for key, value := range properties {
		result[key] = value
	}
	return result
}

func (g *Graph) mergeNodeProperties(node string, properties ...PropertyBag) {
	for _, propertyBag := range properties {
		for key, value := range propertyBag {
			g.nodeProperties[node][key] = value
		}
	}
}

func (g *Graph) mergeEdgeProperties(left, right string, weight float64, properties ...PropertyBag) {
	edgeID := edgeKey(left, right)
	if _, ok := g.edgeProperties[edgeID]; !ok {
		g.edgeProperties[edgeID] = make(PropertyBag)
	}
	for _, propertyBag := range properties {
		for propertyKey, value := range propertyBag {
			g.edgeProperties[edgeID][propertyKey] = value
		}
	}
	g.edgeProperties[edgeID]["weight"] = weight
}

func (g *Graph) setEdgeWeight(left, right string, weight float64) {
	if weight == 0 {
		weight = 1
	}
	if g.repr == AdjacencyList {
		g.adj[left][right] = weight
		g.adj[right][left] = weight
		return
	}
	leftIndex := g.nodeIndex[left]
	rightIndex := g.nodeIndex[right]
	g.matrix[leftIndex][rightIndex] = weight
	g.matrix[rightIndex][leftIndex] = weight
}

func numericValue(value PropertyValue) (float64, bool) {
	switch typed := value.(type) {
	case int:
		return float64(typed), true
	case int8:
		return float64(typed), true
	case int16:
		return float64(typed), true
	case int32:
		return float64(typed), true
	case int64:
		return float64(typed), true
	case uint:
		return float64(typed), true
	case uint8:
		return float64(typed), true
	case uint16:
		return float64(typed), true
	case uint32:
		return float64(typed), true
	case uint64:
		return float64(typed), true
	case float32:
		return float64(typed), true
	case float64:
		return typed, true
	default:
		return 0, false
	}
}
